### PR TITLE
Non-recursive index computation

### PR DESCRIPTION
**Pull request by robrix**
This PR uses an unrolled, templated type family `ElemIndex` & the `KnownNat` class from `GHC.TypeLits` to define membership constraints for unions of up to 150 types.

### CODE REVIEW

File: src/Data/Union.hs [hunk 1]
```diff
@@ -121,17 +126,12 @@ decompose0 (Union _ v) = Right $ unsafeCoerce v
...
-  elemNo :: P t r
-
-instance {-# OVERLAPPING #-} t :< (t ': r) where
-  elemNo = P 0
-
-instance {-# OVERLAPPING #-} t :< r => t :< (t' ': r) where
-  elemNo = P $ 1 + unP (elemNo :: P t r)
-
+elemNo :: forall t r . (t :< r) => P t r
+elemNo = P (fromIntegral (natVal' (proxy# :: Proxy# (ElemIndex t r))))
```
Comment (robrix) [inline src/Data/Union.hs]: I was never comfortable with `elemNo` being defined recursively, “essentially compile-time” or no. Pleasantly, this resolves that ✨

File: src/Data/Union.hs [hunk 2]
```diff
@@ -121,17 +126,12 @@ decompose0 (Union _ v) = Right $ unsafeCoerce v
 weaken :: Union r w -> Union (any ': r) w
 weaken (Union n v) = Union (n+1) v
 
+type (t :< r) = KnownNat (ElemIndex t r)
```
Comment (robrix) [inline src/Data/Union.hs]: Defining `(:<)` as a constraint synonym means that we don’t run into warnings about simplifiable constraints.

File: src/Data/Union.hs (hunk 1 above)
Comment (joshvera) [reply src/Data/Union.hs]: 🤘

Comment (robrix) [reply src/Data/Union.hs]: Made me really happy to finally get an excuse to use `proxy#` _and_ `KnownNat`, too 😆

**Pull request merged**